tests/research/test_batch02_integrity_contracts.py

```
@@ -0,0 +1,367 @@
+        tree = _parse(path)
+        calls = _call_names(tree)
+        function_defs = {
+            node.name for node in ast.walk(tree) if isinstance(node, ast.FunctionDef)
+        }
+
+        assert "_git_sha" not in function_defs, (
+            f"{path.name} may not reintroduce fallback Git identity"
+        )
+        assert "write_text" not in calls and "write_bytes" not in calls, (
+            f"{path.name} may not overwrite evidence with Path.write_*"
+        )
```

Ruff (0.16.2)

[warning] Assertion should be broken down into multiple parts

(PT018)